feat: initial implementation

seixasxbr opened the pull request:
# What is this PR for
Initial implementation of the sonar m750d driver
There is no tests yet since we don't have a sensor sample

# Checklist
- [x] Assign yourself  in the PR

seixasxbr committed:
feat: add Driver

seixasxbr committed:
feat: implement packet extraction

seixasxbr committed:
feat: add method to handle the buffer

seixasxbr committed:
feat: add sonar parse

seixasxbr commented on src/Oculus.h:
This one was copied as is

seixasxbr committed:
feat: add method to process the read buffer

seixasxbr committed:
feat: add fire sonar method

seixasxbr committed:
test: protocol static methods

seixasxbr committed:
feat: define license at manifest

seixasxbr committed:
chore: clear the driver

seixasxbr committed:
chore: clear the protocol

seixasxbr commented on src/Driver.cpp:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    size_t packet_size = header_size + header.payloadSize;
+    if (buffer_size >= packet_size) {
```
```suggestion
    if (buffer_size >= packet_size + payload_size) {
```

get the payload size in the header

doudou replied:
Unless I'm mistaken you'ved one this ? Please remove that comment ?

seixasxbr committed:
chore: clear the code

higor-melo commented on src/Driver.hpp:
```diff
@@ -0,0 +1,28 @@
+        static const int INTERNAL_BUFFER_SIZE = 200000;
+
+        Driver();
+        base::samples::Sonar processOne();
+        void fireSonar(M750DConfiguration const& configuration);
```
Add documentation please

jhonasiv commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,156 @@
+    uint32_t size = 0;
+    uint32_t image_offset = 0;
+
+    if (header.msgVersion == 2) {
```
Very small styling thing, but move the `image_offset = ` to the end of the paragraph so we have a nice sequence of `m_data = ` assignments. A side effect of doing it like you did is that I was looking for some kind of `m_data.<some field> = image_offset`, which obviously doesnt happen.

higor-melo commented on src/M750DConfiguration.hpp:
```diff
@@ -0,0 +1,20 @@
+        double gain = base::unknown<double>();
+        double speed_of_sound = base::unknown<double>();
+        double salinity = base::unknown<double>();
+        bool gain_assist = false;
```
What does `gain assist` do and/or how does it work?

seixasxbr replied:
I am not sure about that yet. There is no doc

jhonasiv commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,156 @@
+
+    if (m_data.bearings) {
+        memcpy(m_data.bearings, buffer + msg_size, ping.nBeams * sizeof(short));
+    }
```
Use your helpers methods
```suggestion
    m_data.image_size = ping_params.imageSize;
    setImage(m_data, ping_params.imageOffset, *buffer);

    m_data.beam_count = ping.nBeams;
    setBearings(m_data, ping.nBeams, *buffer);
```

seixasxbr replied:
I am removing this method, since we won't deal with this kind of message at the moment

higor-melo commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,156 @@
+    uint32_t size = 0;
+    uint32_t image_offset = 0;
+
+    if (header.msgVersion == 2) {
```
When do we have a `msgVersion` that is not 2? Does it happen often or is it a configuration?

seixasxbr replied:
It is a configuration. The simple message could be version 1 or 2

jhonasiv commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,156 @@
+        memcpy(m_data.bearings, buffer + msg_size, ping.nBeams * sizeof(short));
+    }
+
+    m_data.beam_count = ping.nBeams;
```
You would move this assignment upwards due to my suggestion

jhonasiv commented on src/Protocol.hpp:
```diff
@@ -0,0 +1,40 @@
+        SonarData m_data;
+        bool m_simple_ping_result = false;
+
+        static constexpr double HORIZONTAL_FOV_DEG = 130;
+        static constexpr double VERTICAL_FOV_DEG = 20;
```
I would have preferred that this came from a property. We could later rename the repo to `sonar_oculus` and use it for other types of sonars from them, which would have different fovs. For now, keeping it like this is fine, you dont know if we are ever getting a new sonar from them, and if we do it might not even have an equivalent implemetation

doudou replied:
Even more so that these values depend on the current sonar configuration

seixasxbr replied:
<img width="158" height="512" alt="image" src="https://github.com/user-attachments/assets/69c5a09f-bbfb-45fb-a890-7b1f8f68a254" />


This is not the case for this sensor, but I can make it configurable

seixasxbr replied:
done

doudou commented on manifest.xml:
```diff
@@ -3,15 +3,11 @@
   </description>
   <author>Mateus Seixas/mateus.seixas@tidewise.io</author>
-  <license></license>
+  <license>GNU General Public License version 3</license>
```
You really can't be deciding these things alone.

seixasxbr replied:
Ok, sorry.
This is the license of the sdk. I thought that we had to use the same license here, as we based our code in that sdk
<img width="767" height="359" alt="image" src="https://github.com/user-attachments/assets/db5208b7-62ea-4f33-8e23-684aca13ea24" />

doudou commented on src/SonarData.hpp:
```diff
@@ -0,0 +1,20 @@
+        double range = base::unknown<double>();
+        double speed_of_sound = base::unknown<double>();
+        uint8_t* image;
+        short* bearings;
```
**NO**

Use std::vector

doudou commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,156 @@
+
+base::samples::Sonar Protocol::parseSonar()
+{
+    if (!m_simple_ping_result) {
+        throw std::runtime_error("OculusReturnFireMessage parse is not implemented");
+    }
```
Well ... then there should be some check to avoid calling parseSonar if it happens, no ?

Why is it not implemented by the way ?

seixasxbr replied:
I didn't find information about the speed of sound neither about the bearings in this message

doudou commented on src/Driver.hpp:
```diff
@@ -0,0 +1,28 @@
+#ifndef SONAR_OCULUS_M750D_DRIVER_HPP
+#define SONAR_OCULUS_M750D_DRIVER_HPP
+
+#include "M750DConfiguration.hpp"
+#include "Protocol.hpp"
```
`<sonar_oculus_m750d/...>`

doudou commented on src/Driver.cpp:
```diff
@@ -0,0 +1,93 @@
+    writePacket((uint8_t*)&simple_fire_message, sizeof(OculusSimpleFireMessage));
+}
+
+uint8_t setFlags(bool gain_assist)
```
Aren't there anything that could be useful to us here ? I think the range for instance would definitely be ...

seixasxbr replied:
At the moment, I don't think so.
The range is not configurable, this flag is to get the range in meters
<img width="749" height="412" alt="image" src="https://github.com/user-attachments/assets/3899592d-67a5-4cfd-920d-cfc890297595" />
As we get more information about how the sonar works, we can make some of these flags configurable

doudou commented on src/Protocol.hpp:
```diff
@@ -0,0 +1,40 @@
+#ifndef SONAR_OCULUS_M750D_PROTOCOL_HPP
+#define SONAR_OCULUS_M750D_PROTOCOL_HPP
+
+#include "Oculus.h"
```
You cannot have this included in a `.hpp`. It's not namespaced and will pollute every client down the way.

Include it only in the `.cpp`

doudou commented on src/Driver.cpp:
```diff
@@ -0,0 +1,93 @@
+    simple_fire_message.speedOfSound = config.speed_of_sound;
+    simple_fire_message.salinity = config.salinity;
+
+    writePacket((uint8_t*)&simple_fire_message, sizeof(OculusSimpleFireMessage));
```
Use C++ casts, which in this case would be `reinterpret_cast`

doudou commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+void Protocol::handleMessagePingResult(uint8_t const* buffer)
```
If we don't plan to neither have to handle this, please just remove the related code

doudou commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,156 @@
+
+using namespace sonar_oculus_m750d;
+
+void Protocol::handleBuffer(uint8_t const* buffer)
```
This should return a flag saying what got updated, to allow the API client to know whether e.g. there is a new sonar scan to read. We will want to handle configuration at some point, which will break this implementation.

seixasxbr committed:
fix: remove the MessagePingResult handler

there is no information about the bearings neither the speed of sound in this message, so the Sonar struct can't be parsed

seixasxbr committed:
fix: remove license definition

seixasxbr committed:
fix: move oculus.h include to cpp file

seixasxbr committed:
chore: use modern c++ cast

seixasxbr committed:
fix: use vector instead array

seixasxbr committed:
fix: make the beam width and height configurable

seixasxbr committed:
fix: return a bool at buffer handler

seixasxbr committed:
chore: document method

doudou commented on src/Protocol.cpp:
```diff
@@ -0,0 +1,127 @@
+}
+
+void setBearings(SonarData& sonar_data, uint32_t size, uint8_t const* buffer);
+void setImage(SonarData& sonar_data, uint32_t image_offset, uint8_t const* buffer);
```
If these are helper functions, not meant to be available outside this cpp file, declare them as `static`

doudou commented on src/Driver.cpp:
```diff
@@ -0,0 +1,98 @@
+        sizeof(OculusSimpleFireMessage));
+}
+
+uint8_t setFlags(bool gain_assist)
```
`static`

seixasxbr committed:
chore: mark functions as static

seixasxbr committed:
feat: add executable

seixasxbr committed:
fix: bearings parse

seixasxbr committed:
fix: increase internal buffer size

seixasxbr committed:
chore: linting
